esp-hal/src/uart.rs

```diff
@@ -1859,12 +1801,86 @@ mod asynch {
 
         /// Create a new UART instance with defaults in [`Async`] mode.
         pub fn new_async(uart: impl Peripheral<P = T> + 'd, clocks: &Clocks) -> Self {
-            Self::new_async_with_config(
-                uart,
-                Default::default(),
-                None::<TxRxPins<'_, NoPinType, NoPinType>>,
-                clocks,
-            )
+            Self::new_async_with_config(uart, Default::default(), clocks)
+        }
+
+        /// Configure TX and RX pins
+        pub fn with_tx_rx<TX: OutputPin, RX: InputPin>(
+            self,
+            tx: impl Peripheral<P = TX> + 'd,
+            rx: impl Peripheral<P = RX> + 'd,
+        ) -> Self {
+            crate::into_ref!(tx);
+            crate::into_ref!(rx);
+            tx.set_to_push_pull_output(crate::private::Internal);
+            tx.connect_peripheral_to_output(T::tx_signal(), crate::private::Internal);
+
+            rx.set_to_input(crate::private::Internal);
+            rx.connect_input_to_peripheral(T::rx_signal(), crate::private::Internal);
+            self
+        }
+
+        /// Configure TX and RX pins
+        pub fn with_pins<TX: OutputPin, RX: InputPin, CTS: InputPin, RTS: OutputPin>(
+            self,
+            tx: impl Peripheral<P = TX> + 'd,
+            rx: impl Peripheral<P = RX> + 'd,
+            cts: impl Peripheral<P = CTS> + 'd,
+            rts: impl Peripheral<P = RTS> + 'd,
+        ) -> Self {
+            crate::into_ref!(tx);
+            crate::into_ref!(rx);
+            crate::into_ref!(cts);
+            crate::into_ref!(rts);
+
+            tx.set_to_push_pull_output(crate::private::Internal);
+            tx.connect_peripheral_to_output(T::tx_signal(), crate::private::Internal);
+
+            rx.set_to_input(crate::private::Internal);
+            rx.connect_input_to_peripheral(T::rx_signal(), crate::private::Internal);
+
+            cts.set_to_input(crate::private::Internal);
+            cts.connect_input_to_peripheral(T::cts_signal(), crate::private::Internal);
+
+            rts.set_to_push_pull_output(crate::private::Internal);
+            rts.connect_peripheral_to_output(T::rts_signal(), crate::private::Internal);
+            self
+        }
+
+        /// Configure TX pin
+        pub fn with_tx<TX: OutputPin>(self, tx: impl Peripheral<P = TX> + 'd) -> Self {
```

Comment:
It's hard to tell from the diff, so I might not be seeing this correctly, but are these methods duplicated for blocking and async? I think we should be able to have one impl that is generic over the mode.